File: lib/generate-server.ts
import { Param, schemaData } from "./schema";
import { upperFirst, urlToName } from "./utils";

/**
 
type Hello struct {
	Name string `json:"name" binding:"required"`
	Age  int    `json:"age" binding:"min=1,max=120"`
}


 */

const realType = {
  string: "string",
  int: "int64",
  float: "float64",
  bool: "boolean",
  map: "map[string]any",
  any: "any",
  "[]string": "[]string",
  "[]int": "[]int",
  "[]float": "[]float",
  "[]any": "[]any",
  "[]bool": "[]bool",
  "[]map": "[]map[string]any",
};

export function generateServer({ dir, trimUrlName = "" }: { dir: string; trimUrlName?: string }) {
  const list = dir.split("/");
  const dirName = list[list.length - 1];
  let code = "";
  code += `package ${dirName}\n

import "github.com/gin-gonic/gin"\n
  `;

  const apiNameSet = new Set<string>();
  schemaData.forEach(async (schema) => {
    const { url, input, output } = schema;
    const apiName = urlToName(url.replace(trimUrlName, ""));
    if (apiNameSet.has(apiName)) {
      throw new Error(`The ${apiName} entered is already taken. Please choose a different name.`);
    }
    apiNameSet.add(apiName);

    const makeType = (params: Param[], type: "Input" | "Output") => {
      if (!params || !params.length) {
        return;
      }

      code += `
type ${upperFirst(apiName)}${type} struct {\n`;
      let n = 0;
      const subParams = (list: Param[]) => {
        n += 2;
        let space = "";
        for (let i = 0; i < n; i++) {
          space += " ";
        }
        list.forEach((item) => {
          if (item.description) {
            code += `${space}// ${item.description}\n`;
          }
          if (Array.isArray(item.type)) {
            code += `${space}${upperFirst(item.name)} struct {\n`;
            subParams(item.type);
            code += `\n${space}}\n`;
          } else {
            code += `${space}${upperFirst(item.name)} ${realType[item.type]} \`json:"${item.name}" binding:"${
              item.options ? "" : "required"
            }"\`\n`;
          }
        });
      };
      subParams(params);
      code += `\n}\n`;
    };
    makeType(input, "Input");
    makeType(output, "Output");
  });

  code += `type Apis interface {\n`;
  schemaData.forEach(async (schema) => {
    const { description, url } = schema;
    const apiName = urlToName(url.replace(trimUrlName, ""));
    const inputName = `${apiName}Input`;
    const outputName = `${apiName}Output`;
    if (description) {
      code += `// ${description}\n`;
    }
    code += `${apiName}(input *${inputName}) *${outputName}`;
  });
  // Hello(input HelloInput) HelloOutput
  code += "\n}\n";
  code += `func BindHandles(r *gin.Engine, apis Apis) {\n`;
  schemaData.forEach(async (schema) => {
    const { method, url } = schema;
    const apiName = urlToName(url.replace(trimUrlName, ""));
    const inputName = `${apiName}Input`;

    if (method === "GET") {
      code += `r.${method}("${url}", func(c *gin.Context) {
        var input ${inputName}
        if err := c.ShouldBindQuery(&input); err != nil {
          c.JSON(400, map[string]any{"error": err.Error()})
          return
        }
        out := apis.${apiName}(&input)
        c.JSON(200, out)
      })
      `;
    } else {
      code += `r.${method}("${url}", func(c *gin.Context) {
        var input ${inputName}
        if err := c.ShouldBindJSON(&input); err != nil {
          c.JSON(400, map[string]any{"error": err.Error()})
          return
        }
        out := apis.${apiName}(&input)
        c.JSON(200, out)
      })
      `;
    }
  });

  code += `\n}\n`;
  return code;
}

File: old-golang/apigenerate/generateserver.go
package apigenerate

import (
	"fmt"
	"path"
	"strings"

	"project/pkg/strx"
)

var serverRealType = map[ApiType]string{
	"string":   "string",
	"int32":    "int32",
	"int64":    "int64",
	"float":    "float64",
	"bool":     "bool",
	"map":      "map[string]any",
	"any":      "any",
	"[]string": "[]string",
	"[]int":    "[]int32",
	"[]float":  "[]float64",
	"[]any":    "[]any",
	"[]bool":   "[]bool",
	"[]map":    "[]map[string]any",
}

func GenerateServer(opt *GenerateOption) string {
	code := ""
	list := strings.Split(path.Dir(opt.ServerFile), "/")
	dirName := list[len(list)-1]
	code += fmt.Sprintf(`package %v

import "github.com/gin-gonic/gin"
	`, dirName)

	apiNameSet := map[string]any{}
	for _, schema := range schemaData {
		apiName := strx.UrlToName(schema.Url)
		if apiNameSet[apiName] != nil {
			panic(fmt.Sprintf("Api name %v is duplicated", apiName))
		}
		if len(schema.Input) == 0 {
			panic(fmt.Sprintf("need define input: %v", schema.Url))
		}
		if len(schema.Output) == 0 {
			panic(fmt.Sprintf("need define output: %v", schema.Url))
		}
		apiNameSet[apiName] = true
		inputName := apiName + "Input"
		outputName := apiName + "Output"
		makeType := func(params []Param, name string) {
			if len(params) == 0 {
				return
			}
			code += fmt.Sprintf("\ntype %v struct {\n", strx.UpperFirst(name))
			var n int = 0
			var subParams func(list []Param)
			subParams = func(list []Param) {
				n += 2
				space := ""
				for i := 0; i < n; i++ {
					space += " "
				}
				for _, item := range list {
					if item.Description != "" {
						code += fmt.Sprintf("%v// %v \n", space, item.Description)
					}
					if len(item.TypeOfObject) > 0 {
						code += fmt.Sprintf("%v%v struct {\n", space, strx.UpperFirst(item.Name))
						subParams(item.TypeOfObject)
						code += fmt.Sprintf("\n%v}\n", space)
					} else if len(item.TypeOfArrayObject) > 0 {
						code += fmt.Sprintf("%v%v []struct {\n", space, strx.UpperFirst(item.Name))
						subParams(item.TypeOfArrayObject)
						code += fmt.Sprintf("\n%v}\n", space)
					} else {
						code += fmt.Sprintf("%v%v %v `json:\"%v\" binding:\"%v\"` \n", space, strx.UpperFirst(item.Name), serverRealType[item.Type], item.Name, func() string {
							if item.Optional {
								return ""
							}
							return "required"
						}())
					}
				}
			}
			subParams(params)
			code += "\n}\n"
		}
		makeType(schema.Input, inputName)
		makeType(schema.Output, outputName)
	}

	// make Apis
	code += "type Apis interface {\n"
	for _, schema := range schemaData {
		apiName := strx.UrlToName(schema.Url)
		inputName := apiName + "Input"
		outputName := apiName + "Output"
		if schema.Description != "" {
			code += fmt.Sprintf("\n// %v\n", schema.Description)
		}
		code += fmt.Sprintf("%v(input *%v) *%v\n", apiName, inputName, outputName)
	}
	code += "\n}\n"

	// make BindHandles
	code += "func BindHandles(r *gin.Engine, apis Apis) {\n"
	for _, schema := range schemaData {
		apiName := strx.UrlToName(schema.Url)
		realUrl := opt.PrefixUrl + schema.Url
		inputName := apiName + "Input"
		code += fmt.Sprintf(`r.%v("%v", func(c *gin.Context) {
			var input %v
			if err := c.%v(&input); err != nil {
				c.JSON(400, map[string]any{"error": err.Error()})
				return
			}
			out := apis.%v(&input)
			c.JSON(200, out)
		})
				`, schema.Method, realUrl, inputName, func() string {
			if schema.Method == "GET" {
				return "ShouldBindQuery"
			}
			return "ShouldBindJSON"
		}(), apiName)
	}
	code += "\n}\n"

	return code
}
